Use `includes` instead of `indexOf`.

XhmikosR opened the pull request:
This now throws in one case, but I'm not sure if it's a test to blame or if `containsString()` should check for regex and return false explicitly.

Fixes #4414.

XhmikosR commented on lib/utils/containsString.js:
```diff
@@ -34,7 +34,7 @@ function testAgainstString(value, comparison) {
 	}
 
-	if (value.indexOf(comparison) >= 0) {
+	if (value.includes(comparison)) {
```
I thought about this more. I think it would be better to check if `value` isn't a string and return `false` in the first check above, thus dropping the second check.

What do you think?

XhmikosR replied:
Or, check for this in `containsString()`. `!Array.isArray` is true for regex hence why we have this issue.

Not super familiar with the codebase, so let me know which solution you prefer.

hudochenkov replied:
I think we need to `return false` in `testAgainstString()` if `value` is not a string.

XhmikosR replied:
Wouldn't it make more sense to do this in `containsString()` before even calling `testAgainstString()`?

hudochenkov replied:
You're right.

XhmikosR replied:
I'm gonna play with it tomorrow and see how it goes.

XhmikosR commented:
@hudochenkov I'm still not 100% the code is right... The issue comes from

https://github.com/stylelint/stylelint/blob/9071fdd4ef8e2bba78c22fdec54b96f143faed79/lib/rules/comment-word-blacklist/__tests__/index.js#L309

which hits

https://github.com/stylelint/stylelint/blob/9071fdd4ef8e2bba78c22fdec54b96f143faed79/lib/rules/comment-word-blacklist/index.js#L37

Oh, and flow fails in the last patch :)

hudochenkov commented on lib/utils/containsString.js:
```diff
@@ -1,6 +1,8 @@
 /* @flow */
 'use strict';
 
+const _ = require('lodash');
```
You're were removing Lodash... :)

XhmikosR replied:
We can't remove it completely :) No need to reinvent the wheel, unless you want us to maintain our own set of utils :)

hudochenkov commented on lib/utils/containsString.js:
```diff
@@ -28,15 +30,15 @@ module.exports = function containsString(
+		!comparison ||
+		_.isRegExp(comparison) ||
+		(comparison.startsWith('/') && comparison.endsWith('/'))
+	) {
```
Let's split it intro three `if` (like it was before). And use `typeof comparison !== 'string'` instead checking for RegEx.

XhmikosR replied:
Not sure if that will be sufficient, but I will try

hudochenkov commented on lib/utils/containsString.js:
```diff
@@ -28,15 +30,15 @@ module.exports = function containsString(
 		return { match: value, pattern: comparison };
 	}
-
-	return false;
```
I think this one is needed if every condition before fails.

XhmikosR replied:
The way I have the patch, this wouldn't be reached.

XhmikosR commented:
Alright, I made your changes @hudochenkov, but personally I'm not 100% sure it's right. I mean, the typeof check isn't the same as lodash's isRegexp check. Also, we already use this method is many places anyway, hence why I used it in the first place.

BTW, personally I'd merge the checks into one, but maybe keep the end `return false` statement so that flow is happy.

hudochenkov commented:
> I mean, the typeof check isn't the same as lodash's isRegexp check.

This function needs to receive string. So we check if this string or not. We don't want to check if input is not RegExp, is not object, is not array, etc.

Is this PR ready for review?

XhmikosR commented:
> This function needs to receive string. So we check if this string or not. We don't want to check if input is not RegExp, is not object, is not array, etc.

Yeah, but lodash's isString() checks for more:

https://github.com/lodash/lodash/blob/cefddab1cab49189b2ff4d72acf8df7ec723dc22/isString.js#L20

If you think that's OK in this case, we can use the typeof check. Otherwise, there's no point in risking breaking stuff just for this. Remember, lodash is already loaded anyway, and:

1. is used in many places across the codebase (last time I checked we use ~30 functions from lodash)
2. we can't really replace it completely without reinventing the wheel and IMO there's no point in doing so

hudochenkov commented:
You can use `if (!_.isString(comparison)) return false;`.

XhmikosR commented:
There's still a lot of instances that could be replaced with either includes or startsWith, so I'm not sure if this is ready. I could use some help to speed things up.

hudochenkov commented:
We don't really need to convert everything. You did a great job already.

XhmikosR commented:
Alright, I changed most instances in tests too. The remaining ones should be tackled later if ever, and replaced with startsWith/includes if it makes sense.

hudochenkov commented on lib/__tests__/ignoreDisables.test.js:
```diff
@@ -61,10 +61,10 @@ describe('ignoreDisables with standalone', () => {
 
 	it('expected rule 1', () => {
-		expect(results[0].warnings[1].text.indexOf('block-no-empty')).not.toBe(1);
+		expect(results[0].warnings[1].text.includes('block-no-empty')).not.toBe(1);
```
Looks like wrong comparison.

XhmikosR replied:
Yup, I will revert these two, thanks for catching this.

XhmikosR replied:
BTW really nice to see eslint to switch `foo.includes(string).toBe(true)` to `foo.contains(string)` :)

XhmikosR commented on lib/__tests__/defaultSeverity.test.js:
```diff
@@ -14,7 +14,7 @@ it('`defaultSeverity` option set to warning', () => {
 
 		expect(warnings).toHaveLength(1);
-		expect(warnings[0].text.indexOf('block-no-empty')).not.toBe(-1);
+		expect(warnings[0].text).toContain('block-no-empty');
```
BTW my initial patch used `includes`, but ESLint converts those `toContain`
